main.rb

```diff
@@ -1,17 +1,160 @@
+require 'byebug'
+
 module Enumerable
-    def my_each
-        for i in 0...self.size
-            yield(self[i])
-        end
-    self
+  # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+
+  def my_each
+    i = 0
+    if block_given?
+      while i < length
+        yield(self[i])
+        i += 1
+      end
+      return self
+    end
+    to_enum
+  end
+
+  def my_each_with_index
+    i = 0
+    if block_given?
+      while i <= length - 1
+        yield(self[i], i)
+        i += 1
+      end
+      return self
+    end
+    to_enum
+  end
+
+  def my_select
+    arr = []
+    return to_enum unless block_given?
+
+    my_each { |x| arr << x if yield(x) == true }
+    arr
+  end
+
+  def my_all?(arg = nil, &prc)
+    if block_given?
+      my_each { |x| return false if prc.call(x) == false }
+    elsif arg.is_a?(Regexp)
+      my_each { |x| return false if arg.match?(x.to_s) == false }
+    elsif arg.is_a?(Class)
+      my_each { |x| return false if x.is_a?(arg) == false }
+    elsif arg.nil? == false
+      my_each { |x| return false if x != arg }
+    else
+      my_each { |x| return false unless x }
+    end
+    true
+  end
+
+  def my_any?(arg = nil, &prc)
+    if block_given?
+      my_each { |x| return true if prc.call(x) }
+    elsif arg.is_a?(Regexp)
+      my_each { |x| return true if arg.match?(x.to_s) == true }
+    elsif arg.is_a?(Class)
+      my_each { |x| return true if x.is_a?(arg) }
+    elsif arg.nil? == false
+      my_each { |x| return true if x == arg }
+    else
+      my_each { |x| return true if x }
+    end
+    false
+  end
+
+  def my_none?(arg = nil, &prc)
+    if block_given?
+      my_each { |x| return false if prc.call(x) == true }
+    elsif arg.is_a?(Regexp)
+      my_each { |x| return false if arg.match?(x.to_s) == true }
+    elsif arg.is_a?(Class)
+      my_each { |x| return false if x.is_a?(arg) == true }
+    elsif arg.nil? == false
+      my_each { |x| return false if x == arg }
+    else
+      my_each { |x| return false if x.nil? == false }
+    end
+    true
+  end
+
+  def my_count(arg = nil, &prc)
+    count = 0
+    if block_given?
+      my_each { |x| count += 1 if prc.call(x) == true }
+    elsif block_given? == false && arg.nil?
+      my_each { |x| count += 1 if x.nil? == false }
+    elsif arg.nil? == false
+      my_each { |x| count += 1 if x == arg }
     end
+    count
+  end
 
-    def my_each_with_index
-        for i in 0...self.size
-            yield(self[i], i)
+  def my_map(&prc)
+    return to_enum unless block_given?
+
+    arr = []
+    my_each { |x| arr << prc.call(x) }
+    arr
+  end
+
+  def my_inject(arg = nil, arg_2 = nil, &prc)
+    array = to_a
+    ind = 0
+    if block_given? == false
+      if arg.is_a?(Symbol)
+        while ind < array.size - 1
+          array[0] = array[0].send(arg, array[ind + 1])
+          ind += 1
+        end
+      elsif arg.is_a?(Integer)
+        while ind < array.size
```

Comment:
- my_inject. Checkout the [documentation](https://apidock.com/ruby/Enumerable/inject)

- [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block.
![Screenshot from 2020-04-22 08-47-15](https://user-images.githubusercontent.com/30977845/79978198-e5395000-8475-11ea-94a0-e0bf4655543b.png)

- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method
![Screenshot from 2020-04-22 08-50-33](https://user-images.githubusercontent.com/30977845/79978473-5d077a80-8476-11ea-98f2-04c351ee2106.png)